add code/dockerfile for building the secureboot signing job image

chr15p opened the pull request:
This commit adds a stand alone binary (signimage.go) and an associated Dockerfile that together will provide the image for use in the signing job.

The dockerfile compiles signimage, and retrieves the sign-file binary from the kernel-devel package. sign-file is the binary provided by the kernel developers for signing kmods, and while its operation is not super-complex and could be handled within signimage.go, we use it to ensure maximum compatibility. 

The signimage.go binary takes an image name, pulls it down, loops through each file in each layer and if its a kernel modules extracts it to a temporary directory, and runs sign-file on it. All signed kmods are then added back as a (single) new layer, and the new image pushed back to the registry.

It is intended that this output image is the driver container named in km.containerImage

ybettan commented on signimage/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+FROM registry.fedoraproject.org/fedora as ksource
+RUN yum install -y kernel-devel
+RUN cp /usr/src/kernels/*/scripts/sign-file /sign-file
```
This layer seems redundant.

In addition to that, are you assuming anything regarding `/usr/src/kernels/*/scripts/sign-file`?
What happens if there are multiple kernels? are we going to copy multiple files? (When I run that container on my fedora host I don't have any kernel in `/usr/src/kernels/`.

ybettan replied:
Thanks for the explanation.

Regarding the first part, I mean that you have 2 identical layers, so we can combine them (2 layers instead of 3 in total).
You override it so I don't have the original file to see.

chr15p replied:
you're right, that `COPY  cp /usr/src/kernels/*/scripts/sign-file /sign-file` is redundant, I'd not spotted that. I have removed it.

ybettan commented on signimage/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+FROM registry.fedoraproject.org/fedora
+
+COPY --from=builder /workspace/signimage /
+COPY --from=ksource /sign-file /sign-file
```
Any reason we are not using the binary from that layer instead of coping it from the previous identical layer?

chr15p replied:
we could install kernel-devel in this layer, but its huge (80MB plus dependencies) and we only need sign-file (15760 bytes) from it, so we do it this way to try and keep the resultant image as small as possible.

yevgeny-shnaidman replied:
can we just use any image, and not fedora? for example alpine?

chr15p replied:
its a little trickier than that, in that sign-file needs libcrypto, glibc etc, whiich is why the first and third layers are the same base image. In theory we could move to alpine for both,  I'm using fedora because it avoid a bunch of repo and entitlement related issues I had when using ubi-minimal. I dont have enough knowledge of alpine to get sign-image and its deps installed reliably, but theres nothing to stop it being changed in the future.

ybettan commented on signimage/Makefile:
```diff
@@ -0,0 +1,34 @@
+IMAGE_TAG_BASE ?= quay.io/chrisp262/kmod-signer
```
Do we plan to refer to your repo?

qbarrand replied:
I don't think we need new Makefile targets. [`cloudbuild.yaml`](https://github.com/kubernetes-sigs/kernel-module-management/blob/main/cloudbuild.yaml), where we will be adding instructions for this new image, already doesn't use them.

chr15p replied:
The Makefile has been merged into the main Makefile so marking as resolved

ybettan commented on signimage/Makefile:
```diff
@@ -0,0 +1,34 @@
+IMAGE_TAG_BASE ?= quay.io/chrisp262/kmod-signer
+IMAGE_TAG ?= $(shell  git log --format="%H" -n 1)
```
Why are we using a commit sha as the tag instead of always using `latest` for example?

qbarrand replied:
That will also be automated in the postsubmit job. We get both `latest` and the SHA.

chr15p replied:
Again the Makefile is about making debugging easier so it uses the git commit because you probably want latest to reference the latest production version. As Quentin says kmmo itself should never reference it but should use the official image.

ybettan commented on signimage/Makefile:
```diff
@@ -0,0 +1,34 @@
+endif
+
+.PHONY: all
+all: signimage
```
```suggestion
all: sign-image
```

ybettan commented on signimage/Makefile:
```diff
@@ -0,0 +1,34 @@
+	go vet ./...
+
+.PHONY:  signimage
+sign-image: ## Build sign-image binary.
```
```suggestion
signimage: ## Build sign-image binary.
```
Are you OK for having the same name in the target and in the PHONY?

chr15p replied:
yes but I'm not sure that rule should be PHONY at all, it does produce a file.... I'll have a think and fix it either way..

ybettan commented on signimage/README.md:
```diff
@@ -0,0 +1,50 @@
+
+It operates as a wrapper around the sign-file binary provided as part of the kernel-devel package (a wrapper rather than a reimplementation to ensure its bug-for-bug compatabile, rather then having awhole new set of bugs of its own). sign-file is distributed as part of the kernel source so in theory is kernel version specific but in reality it hasn't changed to 5+ years, and for kernel modules to ...
+
+Configuration is done via command line switches or failing that via environment variables
```
Why do we need both switches and env variables? Assuming that KMM's code is the only one consuming that binary, won't it be easier to maintain a single method of passing arguments?

qbarrand replied:
> Why do we need both switches and env variables? Assuming that KMM's code is the only one consuming that binary, won't it be easier to maintain a single method of passing arguments?

Agreed.

chr15p replied:
--help works and produces a usage message, its provided automagically by the flag package.

I'm not assuming this is only used by KMM I think it should at least be possible to use it standalone in some non-KMM build pipeline and supporting environment variables only really adds a couple of lines and actually lets you create defaults taht you can then override with args, so I think the flexability is worth it.

chr15p replied:
If we're only going for one then switches makes more sense, I just like the envars as fallback options. However this is not red line for me, and the discussion is taking more effort than they're worth so I've removed them.  If anyone cares they can always be added back in later.

ybettan commented on signimage/README.md:
````diff
@@ -0,0 +1,50 @@
+
+Or the from the command line
+```
+./signimage -unsignedimage quay.io/chrisp262/minimal-driver:procfsv1 \
````
Should we use a generic image, for example `quay.io/<org>/<image>:<tag>`

ybettan commented on signimage/go.mod:
```diff
@@ -0,0 +1,9 @@
+module readkmod
```
Any special reason this is not called `signimage`?

chr15p replied:
no just a left over from an older name, but this go.mod hass been removed so fixed by default :)

ybettan commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+	// if we've made it this far the arguements are sane
+
+	// get a temp dir to copy kmods into for signing
+	extractiondir, err := os.MkdirTemp("/tmp/", "kmod_signer")
```
From the method's doc:
```
It is the caller's responsibility to remove the directory when it is no longer needed.
```

Should we add a `defer` closing function?

chr15p replied:
its running in a containers ephemeral layer so it will get destroyed when that does, but it's always good to tidy up...  I will add it

qbarrand replied:
Yes, it might be useful during development where `/tmp` has a slightly longer lifetime.

ybettan commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+	// get a temp dir to copy kmods into for signing
+	extractiondir, err := os.MkdirTemp("/tmp/", "kmod_signer")
+	if err != nil {
+		fmt.Errorf("could not create temp dir: %v\n", err)
+		panic(err)
```
```suggestion
		panic(fmt.Errorf("could not create temp dir: %w", err))
```

qbarrand commented on signimage/README.md:
````diff
@@ -0,0 +1,50 @@
+
+```
+Usage of signimage:
+  -cert string
+        path to file containing public key for signing
+  -filestosign string
+        colon seperated list of kmods to sign
+  -key string
+        path to file containing private key for signing
+  -pullsecret string
+        path to file containing credentials for pulling images
+  -pushsecret string
+        path to file containing credentials for pushing images (defaults to the pushsecret)
+  -signedimage string
+        name of the signed image to produce (defaults to "${unsignedimage}-signed")
+  -unsignedimage string
+        name of the image to sign
````
Those work as well with `flag`.

ybettan commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+		kmodstosign[x] = "not found"
+	}
+
+	// set up image download otions
```
```suggestion
	// set up image download options
```

ybettan commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+	}
+
+	// set up image download otions
+	opts := make([]name.Option, 0)
+	ref, err := name.ParseReference(unsignedimagename, opts...)
```
```suggestion
	ref, err := name.ParseReference(unsignedimagename, []name.Option{})
```

I think it is clearer to make sure no one else is using `opts`.

ybettan commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+
+	if nopush == false {
+		// write the image back to the name:tag set via the envvars
+		signedref, err := name.ParseReference(signedimagename, opts...)
```
```suggestion
		signedref, err := name.ParseReference(signedimagename, []name.Option{})
```

I think it is clearer to make sure no one else is using `opts`.

qbarrand commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+package main
```
I think we need unit tests for this software.

ybettan commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+	opts := make([]name.Option, 0)
+	ref, err := name.ParseReference(unsignedimagename, opts...)
+	if err != nil {
+		fmt.Errorf("could not parse the container image name: %v\n", err)
+		panic(err)
```
What does that line do? it is just returning a new error but no one is using it. Am I missing something?
```suggestion
		panic(fmt.Errorf("could not parse the container image name: %w", err))
```

qbarrand replied:
We should not use `panic` and return errors instead.

ybettan commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+
+	a, err := getAuthFromFile(pullsecret, strings.Split(unsignedimagename, "/")[0])
+	if err != nil {
+		fmt.Printf("failed to get auth: %v\n", err)
+		panic(err)
```
Same with the error.

ybettan commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+		te := &transport.Error{}
+
+		if errors.As(err, &te) && te.StatusCode == http.StatusNotFound {
+			fmt.Errorf("could not get image: http.StatusNotFound")
```
Who uses the return value of that?

ybettan commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+			fmt.Errorf("could not get image: http.StatusNotFound")
+		}
+
+		fmt.Errorf("could not get image: %v", err)
```
Who uses the return value of that?

ybettan commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+	fmt.Printf("\n== Looking for files: %s\n", strings.Replace(fileslist, ":", " ", -1))
+
+	/*
+	** loop through all the layers in the image from the top down
```
Why do we need to loop over all the layers? Isn't inspecting the last layer enough?

chr15p replied:
that depends on if you've put all your kmods in the last layer :)
kvc-simple-kmod for example uses two seperate COPY statements so puts its two kmods in two different layers

ybettan commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+	 */
+	layers, _ := img.Layers()
+	for i := len(layers) - 1; i >= 0; i-- {
+		fmt.Printf("== Searching layer %d\n", i)
```
Should we use logs?

ybettan commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+		currentlayer := layers[i]
+		layerreader, err := currentlayer.Uncompressed()
+		if err != nil {
+			fmt.Errorf("could not get layer: %v", err)
```
Who uses the return value?

qbarrand commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+		panic(err)
+	}
+	fmt.Printf("\n== Successfully pulled image: %s\n", unsignedimagename)
+	fmt.Printf("\n== Looking for files: %s\n", strings.Replace(fileslist, ":", " ", -1))
```
Could we use a logging framework here? [logr](https://github.com/go-logr/logr) as the facade and [klog](https://github.com/kubernetes/klog) as the implementation work well in the community.

chr15p replied:
We could but I don't see what benefit it would add, this is a pretty simple batch job, that runs in its own container, and then exits. It's doing one thing and (hopefully) doing it well.

The complexity of using a full blown logging framework just so I can dump a handful of progress lines to stdout feels excessive to me, great for long running stuff like controllers, overkill for this.

qbarrand replied:
The complexity of setting up `klog` together with `logr` is moderate: https://github.com/kubernetes-sigs/kernel-module-management/blob/main/main.go#L78-L82
It would give us fields, levelled logging, the ability to mock the logger in unit tests, and more.
This is also the [preferred approach in Kubernetes](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#logging-in-kubernetes):
> Kubernetes project uses [klog](https://github.com/kubernetes/klog) for logging.

qbarrand commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+		 */
+		tarreader := tar.NewReader(layerreader)
+		for {
+			header, _ := tarreader.Next()
```
We should not ignore the error here.

qbarrand commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+	}
+
+	//turn our tar archive into a layer
+	mediatype, _ := layers[len(layers)-1].MediaType()
```
We should not ignore the error here.

yevgeny-shnaidman commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+		panic(err)
+	}
+
+	descriptor, err := remote.Get(ref, remote.WithAuth(a))
```
We already have a registry package that has API for accessing the image, get layers of the image and checking if a file is present in the image. Maybe we can use it ( with some additional APIs). This will probably save a lot of code lines that you have here

chr15p replied:
I disagree.

Firstly:  This is a standalone piece of code intended to run in a job that (may or may not be) run from KMMO. It is really only is in this repo at all because it need to be somewhere, and it not being in my personal github repo makes some external political things easier. The idea of closely coupling this code with that of the controller seems unnecessary. 

Secondly: After a quick look, while Registry superficially works in the same space, it would need significant changes to it make it usable, anything that references kmmv1beta1 would need to be removed, signimage does not have access to the crd, nor does it need it, but Registry refers to it extensively. 

Thirdly: Checking if a single image is present in the top layer is not enough here, we need to find all kmods (issue #94), in all layers (the code to loop through layers isn't in Registry), change them (not in Registry), add them to a new layer (not in Registry), write the image back (not in Registry). So there's a lot of functionality registry is missing we'd need to add.

So as it stands there is not much overlap between my code and Registry so it would need to be pretty much rewritten from scratch.

yevgeny-shnaidman replied:
actually, i think that the work to be done is not that big:
1) since your code is in this repo, why not the already available package? If, sometime in the future,  we decide that your code needs to be removed to a different repo, we can simply make registry a public package
2) registry is only using the PullOption struct of the kmmv1beta, it does not receive a CRD as a parameter anywhere. You will need to configure your access to the registry, so why not use kmmv1beta structs, as opposed to using structs defined in other packages?
3) we will deal with finding all kmod later ( your code right now also does not handle it). You main function can be describe as:

-  get the image
-  get all layers of image
-  loop over each layer and check if a file exist
-  in case file exists - get it, sign and add it
So we can add extraction function to the package and also add function, and use the existing function. I think is will significantly reduce your code and enhance our package, which is a good thing. The loop can stay in the main, just as it is not in preflight package, for example

chr15p replied:
> we will deal with finding all kmod later ( your code right now also does not handle it).

yes it does, somewhere around line 284 
`if kmodstosign[canonfilename] == "not found" || (fileslist == "" && kmodstosign[canonfilename] == "") {`

this checks if the file has been passed in via filelist (that then populates kmodstosign),  or if filelist is empty and we haven't seen this kmod before, then sign anyway. 

(this is why Sign.FilesToSign in the api was set as optional in the crd :) )

I'll have a look at the rest of it tomorrow

chr15p replied:
Lots of the generic code has been moved into internal/registry/registry.go and the binary has moved into cmd/signimage so tI think that resolves (or at least outdates) this

ybettan commented:
/ok-to-test

chr15p commented on signimage/README.md:
```diff
@@ -0,0 +1,50 @@
+	-pullsecret pullsecret \
+	-key mykey.priv \
+	-cert mykey.der \
+	-filestosign /modules/simple-kmod.ko:/modules/simple-procfs-kmod.ko
```
if the image name is generic these should probably be generic as well rather than simple-kmod or kmm-mod e.g </var/lib/kmod1.ko>[:</var/lib/kmod2.ko>]...

ybettan replied:
Indeed.

chr15p commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+		fmt.Errorf("could not Image(): %v", err)
+		panic(err)
+	}
+	fmt.Printf("\n== Successfully pulled image: %s\n", unsignedimagename)
+	fmt.Printf("\n== Looking for files: %s\n", strings.Replace(fileslist, ":", " ", -1))
```
no, its running in its own container so stdout is being captured by k8s automatically so these Printf statements are available via kubectl logs <pod>, messages from this don't end up in the KMM controller logs so dont need the same format

yevgeny-shnaidman commented on signimage/signimage.go:
```diff
@@ -0,0 +1,352 @@
+
+	//make a map of the files to sign so we can track what we want to sign
+	kmodstosign := make(map[string]string)
+	for _, x := range strings.Split(fileslist, ":") {
```
how do we expect to get the list of .ko files to sign? Will we expect only the files' names (e.i. simple-kmod.ko) or the whole path ( /lib/modules/kernel/simple-kmod.ko)?

chr15p replied:
we're expecting the full paths with colon separators, e.g.  
`-filestosign /opt/lib/modules/4.18.0-348.2.1.el8_5.x86_64/simple-kmod.ko:/opt/lib/modules/4.18.0-348.2.1.el8_5.x86_64/simple-procfs-kmod.ko`
we could change it to be just the filename but I'm not sure that really adds anything, and prevents the (admittedly edge) use case of a single signed-image supporting multiple versions of the same kmod each signed for a different key.

yevgeny-shnaidman commented on signimage/signimage.go:
```diff
@@ -0,0 +1,352 @@
+			// paths in a layer are relative, and supplied paths are absolute so canonicalise
+			canonfilename := canonicalisePath(header.Name)
+			//either the kmod has not yet been found, or we didn't define a list to search for
+			if kmodstosign[canonfilename] == "not found" || (fileslist == "" && kmodstosign[canonfilename] == "") {
```
i might be missing something, but it seems that in case of an empty filelist, this code will be signing every file/directory in the layer,no?

chr15p replied:
yep, that was a bug that crept in while I was debugging something else :)  should be fixed now

yevgeny-shnaidman commented on signimage/signimage.go:
```diff
@@ -0,0 +1,352 @@
+}
+
+func canonicalisePath(path string) string {
+	canonical := strings.Replace("/"+path, "/./", "/", -1)
```
since the files in the layer are in relative path, isn't it enough just to add a "/" to the start of the path of the file?

chr15p replied:
The files in a layer are normally relative but I think their exact format depends how exactly they were created. They are sometimes stored as  'path/` and sometimes `./path/` and sometimes aren't relative at all but are absolute (I suspect this is in baselayers) which become `//path` with a "/" appended. 

In the CRD we are asking for absolute paths, so this function is just doing some search and replace to clean up those patterns which then allows us to do string comparison on the paths to easily check if the file we have in the image is one to sign

qbarrand commented on signimage/signimage.go:
```diff
@@ -0,0 +1,352 @@
+	dirname := filepath.Dir(header.Name)
+	err := os.MkdirAll(tmpdir+"/"+dirname, 0750)
+	if err != nil {
+		die(2, "could not create tempdir for kmod", err)
```
I would not call `os.Exit` from anywhere except the `main` function. The rest of the code should return errors to be processed by `main`.

chr15p replied:
I'm leaving the die() helper function calls in main() because to unroll them would result in lots of repetitive logging/exit code, but you're right calling it from many layers down the stack in the image walking code is pretty nasty. Fixed.

chr15p commented:
I've done some major restructuring based on comments.
- moved to using klogr for printing to stdout
- moved a lot of the mainloop functionality to helper functions
- moved those helper functions to a library

I tried to put the helper functions into internal/registry/registry.go but that has a bunch of practical issues, libraries in internal/ are only available to things in directories above them so binaries in signimge/ are prohibited from seeing them. You can work around this for compilation with a bit of Makefile cleverness, but controller-gen (as used in `make generate`) really doesn't like it   (best option I could find to get it to ignore the directory ran into https://github.com/kubernetes-sigs/controller-tools/issues/684 )

So for the moment I have my own registry.go, if we can work around the controller-gen issue  we should be able to merge signimage/registry.go and internal/registry/registry.go easily.

qbarrand commented:
@chr15p having these files as separate Go module make them unable to import the KMM libraries indeed.
Moving `signimage.go` to a directory under `cmd` (such as `cmd/sign/signimage.go` for instance) would eliminate that problem by including the new code into the existing module that we have for KMM. Is there any downside in doing that?

chr15p commented:
I have no issue with moving it somewhere but I still cant get it to work

> [cprocter@toolbox kernel-module-management-upstream]$ ls cmd/signimage/signimage.go                                                                                                                              
cmd/signimage/signimage.go
> [cprocter@toolbox kernel-module-management-upstream]$ go build cmd/signimage/signimage.go                                                                                                                        
> [cprocter@toolbox kernel-module-management-upstream]$ make generate
> go install github.com/golang/mock/mockgen@v1.6.0
> /home/cprocter/engineering/kernel-module-management-upstream/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."                                                                         
> signimage.go:18:2: use of internal package github.com/kubernetes-sigs/kernel-module-management/internal/registry not allowed                                                                                      
> Error: not all generators ran successfully
> run `controller-gen object:headerFile=hack/boilerplate.go.txt paths=./... -w` to see all available markers, or `controller-gen object:headerFile=hack/boilerplate.go.txt paths=./... -h` for usage                
>make: *** [Makefile:99: generate] Error 1

This is similar to the error when I compile signimage.go from within cmd/signimage/   so it feels like controller-gen is recursing into the directory and doing its thing.

I've tried moving it to other places as well and same result :(

yevgeny-shnaidman commented on internal/registry/registry.go:
```diff
@@ -245,3 +259,124 @@ func (r *registry) getImageDigestFromMultiImage(manifestListStream []byte) (stri
+** data[] - an array of strings to pass through to the helper function
+**        to provide any additional data needed.
+ */
+func (r *registry) WalkFilesInImage(image v1.Image, fn func(filename string, header *tar.Header, tarreader io.Reader, data []string) error, data []string) error {
```
This function walk on all the files in all the layers. We are only interested in the file under specific directory. So maybe we should add an option to this function or to processing function. WDYT?

chr15p replied:
KMMO is only interested in the files under specific directory, but I am trying to keep this image as generic as possible so it is still usable outside KMMO. That means we can't assume that all the kmods are in a single directory, so it would have to be an array (or similar) of directories and that adds a lot more complication.  Plus then future users may have different requirements.

The signimage job currently runs through in a few tens of seconds, so I don't really think there's much need to optomise this function for performance at the moment.

So, maybe in the future, but I dont think we need to do that in this PR.

chr15p commented on internal/registry/registry.go:
```diff
@@ -44,6 +52,12 @@ type Registry interface {
 	GetLayersDigests(ctx context.Context, image string, po *kmmv1beta1.PullOptions, registryAuthGetter auth.RegistryAuthGetter) ([]string, *RepoPullConfig, error)
 	GetLayerByDigest(digest string, pullConfig *RepoPullConfig) (v1.Layer, error)
+	WriteImageByName(imageName string, image v1.Image, auth authn.Authenticator) error
+	WalkFilesInImage(image v1.Image, fn func(filename string, header *tar.Header, tarreader io.Reader, data []string) error, data []string) error
```
Done.
The array of strings was an attempt to make this a general purpose function but your way is much cleaner.

chr15p commented on signimage/signimage.go:
```diff
@@ -0,0 +1,357 @@
+
+	var unsignedimagename string
+	var signedimagename string
+	var fileslist string
```
think I have caught all of these let me know if you spot one I've missed...

yevgeny-shnaidman commented on cmd/signimage/signimage.go:
```diff
@@ -0,0 +1,309 @@
+	return
+}
+
+func canonicalisePath(path string) string {
```
```suggestion
func canonicalisePath(path string) string {
       filepath.IsAbs(path) {
           return path
       }
       return "/" + filepath.Clean(path)
}
       
```

yevgeny-shnaidman replied:
Also, lets add a comment that explains all the various filepaths scenarios that can be defined by the layet

chr15p replied:
I've pushed a variant of this (appending the "/" needs to be inside the clean, also I think its good to canonicalise absolutepaths just in case they have "./" etc in them), I've also changed it so the filestosign list gets canonicalised as well, just in case someone doesn't read the instructions and tries to sign something with a relative path, or adds "../" type things to the path.

yevgeny-shnaidman replied:
Kudos for your version, it is shorter :). Regarding filestosign: do we want to try to correct users' mistakes? What if by trying to correct their mistake we will actually sign a file in an incorrect path? I think that in case user has provided incorrect path, we can just print a warning message , and that's it. WDYT?

chr15p replied:
I used to write perl as a job, shorter is not always better!

I agree about chance for signing the wrong file, I think its very unlikely to ever happen but its cleaner just to put out an error message and let the user fix their problem rather than take the risk. I have updated the code.

yevgeny-shnaidman commented on cmd/signimage/signimage.go:
```diff
@@ -0,0 +1,309 @@
+	return strings.Replace(canonical, "//", "/", -1)
+}
+
+func extractFile(tmpdir string, header *tar.Header, tarreader io.Reader) (string, error) {
```
why not make this function part of registry package?

chr15p replied:
We'd end up with signimage calling into registry which calls back into signimage, which in turn calls back into registry, which feels complicated.

But then additionally extractFile is pretty much tailored for the signimage use case (it expects io.Reader to be lined up ready to read the first file, then extracts to a specific place in /tmp that maintains the path information so it can be added to the new layer, and creates things with specific permissions), some of this could be factored out into registry but its not generic enough to all go in there, 

So the code is simple if its in signimage but moving those few lines to registry makes it a lot more complex.

yevgeny-shnaidman replied:
besides persmissions, which can be passed as a parameter, i don't see anything sign-related code in this function. io.Reader can be set to whatever offset user want prior to calling function, and extracting file to user-defined directory also seems ok for general use case

chr15p replied:
Ok, I've moved that functionality to the registry library. I think it makes the signimage code worse because it has to pass the registry object into itself, so it can be passed back to signimage, so it can call functions in registry, but I guess if anyone else ever uses the registry library for extracting files it will help them.

yevgeny-shnaidman commented:
/approve

yevgeny-shnaidman commented:
/lgtm
